NonBayesian/Second/second2.py

```diff
@@ -0,0 +1,226 @@
+           [0., 1., 1., 1., 0.],
+           [0., 0., 0., 0., 0.]])
+    """
+    if arr.ndim == 1:
+        n = np.shape(arr)[0]
+        z_arr = np.zeros(n + 2)
+        z_arr[1:n+1] = arr
+        return z_arr
+
+    if arr.ndim == 2:
+        n = np.shape(arr)[0]
+        m = np.shape(arr)[1]
+        z_arr = np.zeros((n + 2, m+2))
+        z_arr[1:n + 1, 1:m+1] = arr
+        return z_arr
+
+    if arr.ndim == 3:
+        n = np.shape(arr)[0]
+        m = np.shape(arr)[1]
+        k = np.shape(arr)[2]
+        z_arr = np.zeros((n + 2, m + 2, k + 2))
+        z_arr[1:n + 1, 1:m + 1, 1:k + 1] = arr
+        return z_arr
+
+    raise Exception("Bad arguments")
+
+
+def cache(arr):
+    """
+    >>> cache(np.ones(10))
+    array([ 1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10.])
+    >>> cache(np.ones((3, 3)))
+    array([[1., 2., 3.],
+           [2., 4., 6.],
+           [3., 6., 9.]])
+    """
+    cache_arr = np.zeros(np.shape(arr))
+
+    if arr.ndim == 1:
+        for i in range(0, np.shape(arr)[0]):
+            cache_arr[i] = np.sum(arr[0:i+1])
+        return cache_arr
+
+    if arr.ndim == 2:
+        for i in range(0, np.shape(arr)[0]):
+            for j in range(0, np.shape(arr)[1]):
+                cache_arr[i, j] = np.sum(arr[0:i+1, 0:j+1])
+        return cache_arr
+
+    if arr.ndim == 3:
+        for i in range(0, np.shape(arr)[0]):
+            for j in range(0, np.shape(arr)[1]):
+                for k in range(0, np.shape(arr)[2]):
+                    cache_arr[i, j, k] = np.sum(arr[0:i+1, 0:j+1, 0:k+1])
+        return cache_arr
+
+    else:
+        raise Exception("Bad dimension")
+
+
+def cache2(arr):
+    """
+        >>> cache2(np.ones(10))
+        array([ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10.,  0.])
+        >>> cache2(np.ones((3, 3)))
+        array([[0., 0., 0., 0., 0.],
+               [0., 1., 2., 3., 0.],
+               [0., 2., 4., 6., 0.],
+               [0., 3., 6., 9., 0.],
+               [0., 0., 0., 0., 0.]])
+
+    """
+    if arr.ndim == 1:
+        n = np.shape(arr)[0]
+        cache_arr = np.zeros(n + 2)
+        for i in range(1, n + 1):
+            cache_arr[i] = cache_arr[i-1] + arr[i-1]
+        return cache_arr
+    if arr.ndim == 2:
+        n = np.shape(arr)[0]
+        m = np.shape(arr)[1]
+        cache_arr = np.zeros((n + 2, m + 2))
+        for i in range(1, n + 1):
+            for j in range(1, m + 1):
+                cache_arr[i, j] = cache_arr[i, j - 1] + cache_arr[i - 1, j] - cache_arr[i - 1, j - 1] + arr[i - 1, j - 1]
+        return cache_arr
+
+    if arr.ndim == 3:
+        n = np.shape(arr)[0]
+        m = np.shape(arr)[1]
+        k = np.shape(arr)[2]
+        cache_arr = np.zeros((n + 2, m + 2, k + 2))
+        for i in range(1, n + 1):
+            for j in range(1, m + 1):
+                for k in range(1, k + 1):
+                    cache_arr[i, j, k] = cache_arr[i-1, j ,k] + cache_arr[i, j-1, k] + cache_arr[i, j, k-1] \
```

Comment:
Use parentheses. Don't use these slashes
```suggestion
                    cache_arr[i, j, k] = (
                        cache_arr[i-1, j ,k] + cache_arr[i, j-1, k] + cache_arr[i, j, k-1]
```